main.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2020 The cert-manager authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"os"
+	"time"
+
+	capi "k8s.io/api/certificates/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	// +kubebuilder:scaffold:imports
+
+	"github.com/cert-manager/signer/controllers"
+	"github.com/cert-manager/signer/internal/kubernetes/signer"
+)
+
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	_ = capi.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
+}
+
+func main() {
+	var metricsAddr string
+	var enableLeaderElection bool
+	var signerName string
+	var caCertPath string
+	var caKeyPath string
+	var certificateDuration time.Duration
+
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
+	flag.StringVar(&signerName, "signer-name", "example.com/foo", "Only sign CSR with this .spec.signerName.")
+	flag.StringVar(&caCertPath, "ca-cert-path", "ca.pem", "Sign CSR with this certificate file.")
+	flag.StringVar(&caKeyPath, "ca-key-path", "ca-key.pem", "Sign CSR with this private key file.")
+	flag.DurationVar(&certificateDuration, "certificate-duration", time.Hour, "The duration of the certificate signature.")
+	flag.Parse()
+
+	ctrl.SetLogger(zap.New(zap.UseDevMode(true)))
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: metricsAddr,
+		Port:               9443,
+		LeaderElection:     enableLeaderElection,
+		LeaderElectionID:   "c44f5c85.cert-manager.io",
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	signer, err := signer.NewSigner(caCertPath, caKeyPath, certificateDuration)
+	if err != nil {
+		setupLog.Error(err, "unable to start signer")
+		os.Exit(1)
+	}
```

Comment:
Hm, would it be worth moving this call into the `SetupWithManager` function and then exporting `CAFile string` and `CAPrivateKeyFile string` as fields in the reconciler, and *not* exporting the `signer` itself? Bit of a nit though :) (we could also do this with _more_ things too).

Generally I think packages that expose a `New` function that either takes an Options struct or just a load of arguments works out 'cleaner' in Go for this sort of thing, as it avoids ever allowing users to create a struct that is 'half initialized' (i.e. some values set but Setup() has not been called). Not too fussed and not too important in this context though so 🤷‍♂

Reply:
I agree and if we decide to take this experiment further I'll do this. For now I'll leave it.